File: src/hash.rs
use crate::domain::DomainSeparator;
use crate::keccak::keccak256;
use crate::traits::Eip712Struct;

/// Compute the final EIP-712 typed data digest.
pub fn hash_typed_data(domain: &DomainSeparator, message: &impl Eip712Struct) -> [u8; 32] {
    let mut data = [0u8; 66];
    data[0] = 0x19;
    data[1] = 0x01;
    data[2..34].copy_from_slice(&domain.separator_hash());
    data[34..66].copy_from_slice(&message.hash_struct());
    keccak256(&data)
}

#[cfg(test)]
mod tests {
    extern crate alloc;

    use alloc::vec::Vec;

    use super::*;
    use crate::domain::DomainBuilder;
    use crate::keccak::keccak256;
    use crate::traits::Eip712Struct;

    struct SimpleStruct {
        value: [u8; 32],
    }

    impl Eip712Struct for SimpleStruct {
        fn type_string() -> &'static str {
            "SimpleStruct(uint256 value)"
        }

        fn type_hash() -> [u8; 32] {
            keccak256(Self::type_string().as_bytes())
        }

        fn encode_data(&self) -> Vec<u8> {
            self.value.to_vec()
        }
    }

    #[test]
    fn test_hash_typed_data_format() {
        let domain = DomainBuilder::new().name("Test").version("1").build();
        let msg = SimpleStruct { value: [0x01; 32] };
        let hash = hash_typed_data(&domain, &msg);

        let mut expected_input = Vec::new();
        expected_input.push(0x19);
        expected_input.push(0x01);
        expected_input.extend_from_slice(&domain.separator_hash());
        expected_input.extend_from_slice(&msg.hash_struct());
        let expected = keccak256(&expected_input);

        assert_eq!(hash, expected);
    }
}

File: go/hash.go
package eip712

import "fmt"

// HashStruct computes the EIP-712 struct hash for a typed message.
//
//	keccak256(typeHash || encodeData(field1) || encodeData(field2) || ...)
func HashStruct(primaryType string, types TypeDefinitions, message map[string]interface{}) ([32]byte, error) {
	fields, ok := types[primaryType]
	if !ok {
		return [32]byte{}, fmt.Errorf("eip712: type %q not found in TypeDefinitions", primaryType)
	}

	typeStr, err := BuildCanonicalTypeString(primaryType, types)
	if err != nil {
		return [32]byte{}, err
	}
	typeHash := ComputeTypeHash(typeStr)

	encoded := make([]byte, 0, 32*(1+len(fields)))
	encoded = append(encoded, typeHash[:]...)

	for _, field := range fields {
		value := message[field.Name]
		slot, err := EncodeField(field.Type, value, types)
		if err != nil {
			return [32]byte{}, fmt.Errorf("eip712: encoding field %q (%s): %w", field.Name, field.Type, err)
		}
		encoded = append(encoded, slot[:]...)
	}

	return Keccak256(encoded), nil
}

// HashTypedData computes the full EIP-712 digest.
//
//	keccak256(0x19 || 0x01 || domainSeparator || structHash)
func HashTypedData(
	domain EIP712Domain,
	types TypeDefinitions,
	primaryType string,
	message map[string]interface{},
	opts *TypedDataOptions,
) ([32]byte, error) {
	domainSep, err := HashDomainSeparator(domain, opts)
	if err != nil {
		return [32]byte{}, fmt.Errorf("eip712: domain separator: %w", err)
	}

	structHash, err := HashStruct(primaryType, types, message)
	if err != nil {
		return [32]byte{}, fmt.Errorf("eip712: struct hash: %w", err)
	}

	return assembleDigest(domainSep, structHash), nil
}

// HashTypedDataRaw is the low-level variant accepting pre-computed inputs.
// typeHash is the 32-byte type hash; encodedStruct is the ABI-encoded struct
// fields WITHOUT the leading typeHash (HashStruct prepends it internally).
func HashTypedDataRaw(
	domain EIP712Domain,
	typeHash [32]byte,
	encodedStruct []byte,
	opts *TypedDataOptions,
) ([32]byte, error) {
	domainSep, err := HashDomainSeparator(domain, opts)
	if err != nil {
		return [32]byte{}, fmt.Errorf("eip712: domain separator: %w", err)
	}

	// Struct hash = keccak256(typeHash || encodedStruct)
	raw := make([]byte, 0, 32+len(encodedStruct))
	raw = append(raw, typeHash[:]...)
	raw = append(raw, encodedStruct...)
	structHash := Keccak256(raw)

	return assembleDigest(domainSep, structHash), nil
}

// assembleDigest builds the EIP-191 encoded digest.
// Format: keccak256(0x19 || 0x01 || domainSeparator || structHash)
func assembleDigest(domainSep, structHash [32]byte) [32]byte {
	buf := make([]byte, 66)
	buf[0] = 0x19
	buf[1] = 0x01
	copy(buf[2:34], domainSep[:])
	copy(buf[34:66], structHash[:])
	return Keccak256(buf)
}
